### DISCUSSION TITLE
[RAC] `Namespace 'React' has no exported member 'JSX'`

### DISCUSSION THREAD

**Discussion by mgo-af** [adobe/react-spectrum#4735] [Q&A]
Hello everyone 

Our design system is shipped via a monorepo implemented with yarn workspaces and we publish the React components to a private npm registry. That library includes a folder with a simple app that we use to check for type augmentations with `skipLibCheck: true` in its `tsconfig.json` file.

For context:

1. We are composing some components using `react-aria-components`
2. We are using Yarn 3 due to a compatibility limitation between the private npm registry vendor we have to use and Yarn 1. This means we can't use resolutions in nested dependencies to force certain versions of `@types/react`, for example
3. There is one additional package in the monorepo called `/releases` that calls NextJS 12 a dependency, and that depends on the components library, which we instruct to depend on using `@our-lib/react: "*"`.

Yesterday we started noticing that `tsc --noEmit` displays these errors:

```
Error: node_modules/@react-aria/focus/dist/types.d.ts(51,59): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/i18n/dist/types.d.ts(22,63): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/overlays/dist/types.d.ts(136,65): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/overlays/dist/types.d.ts(156,67): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/overlays/dist/types.d.ts(199,65): error TS2[694](https://github.com/fuegokit/fuegokit/actions/runs/5406483609/jobs/9823410782#step:9:695): Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/select/dist/types.d.ts(115,30): error TS2304: Cannot find name 'TrustedHTML'.
Error: node_modules/@react-aria/select/dist/types.d.ts(303,69): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/@react-aria/visually-hidden/dist/types.d.ts(26,67): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/react-aria-components/dist/types.d.ts(824,65): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/react-aria-components/dist/types.d.ts(962,61): error TS2694: Namespace 'React' has no exported member 'JSX'.
Error: node_modules/react-aria-components/dist/types.d.ts(1638,76): error TS2694: Namespace 'React' has no exported member 'JSX'.
```
So one of the monorepo packages is including some React types that `react-aria` doesn't like, and I'm trying to run them down now.

Additionally, the builds for the Next 12 project fail with the same failures that @LFDanLu observed [here](https://github.com/adobe/react-spectrum/pull/4698/files#r1239089403):

```
➤ YN0000: [@our-lib/releases]: ../node_modules/@internationalized/date/dist/import.mjs
➤ YN0000: [@our-lib/releases]: Module not found: Can't resolve '@swc/helpers/_/_class_private_field_init' in '/home/runner/work/our-lib/our-lib/node_modules/@internationalized/date/dist'
➤ YN0000: [@our-lib/releases]: 
➤ YN0000: [@our-lib/releases]: Import trace for requested module:
➤ YN0000: [@our-lib/releases]: ../node_modules/@internationalized/date/dist/import.mjs
➤ YN0000: [@our-lib/releases]: ../node_modules/react-aria-components/dist/import.mjs
➤ YN0000: [@our-lib/releases]: ../react/lib-esm/components/Combobox/Combobox.js
➤ YN0000: [@our-lib/releases]: ../react/lib-esm/components/Combobox/index.js
➤ YN0000: [@our-lib/releases]: ../react/lib-esm/index.js
➤ YN0000: [@our-lib/releases]: 
➤ YN0000: [@our-lib/releases]: ../node_modules/react-aria/node_modules/@internationalized/date/dist/import.mjs
➤ YN0000: [@our-lib/releases]: Module not found: Can't resolve '@swc/helpers/_/_class_private_field_init' in '/home/runner/work/our-lib/our-lib/node_modules/react-aria/node_modules/@internationalized/date/dist'
➤ YN0000: [@our-lib/releases]: 
➤ YN0000: [@our-lib/releases]: Import trace for requested module:
➤ YN0000: [@our-lib/releases]: ../node_modules/react-aria/node_modules/@internationalized/date/dist/import.mjs
➤ YN0000: [@our-lib/releases]: ../node_modules/react-aria/node_modules/@react-aria/calendar/dist/import.mjs
➤ YN0000: [@our-lib/releases]: ../node_modules/react-aria/dist/import.mjs
➤ YN0000: [@our-lib/releases]: ../react/lib-esm/hooks/index.js
➤ YN0000: [@our-lib/releases]: ../react/lib-esm/index.js
```

I understand that it's hard to opine without a repro, but can anyone make a recommendation for how they might tackle this?

**Comment by LFDanLu** [comment]
Just a fyi with regards to the Next 12 + swc issues issues, https://github.com/vercel/next.js/issues/48593 seems to be where this was discussed at length. Unfortunately the bulk of the discussion seems centered around the issues with Next 13, haven't tried the [overrides solution](https://github.com/vercel/next.js/issues/48593#issuecomment-1522095678) applies here.

**Reply by LFDanLu** [reply]
@mgo-af btw there has been an update on that issue with a way to resolve the swc issues for Next 12: https://github.com/vercel/next.js/issues/48593#issuecomment-1681930868.

**Comment by alanhajo-ui** [comment]
Having the same issue. Not running Nextjs.
Anyone got a solution?

**Reply by alanhajo-ui** [reply]
Created issue:
https://github.com/adobe/react-spectrum/issues/5157
